### PR TITLE
refactor csv export functions into library

### PR DESCRIPTION
**Pull request by jim68000**
This moves the common functions from Stu's CSV export work into a library as preparation for adding export to other entities. As Lee likes screenshots...

image

### CODE REVIEW

File: src/lib/export-helper.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  return userPermissions.includes(targetPermission)
+}
+
+function tooManyItems (resultCount, maxItems) {
+  return resultCount >= maxItems
+}
+
+function buildExportMessage (resultCount, maxItems) {
+  if (tooManyItems(resultCount, maxItems)) {
+    return `Filter to less than ${maxItems} projects to download`
```
Comment (reupen) [inline src/lib/export-helper.js]: Shouldn't the correct noun (projects) be passed in as well?

Comment (jim68000) [reply src/lib/export-helper.js]: Good spot, well fix

File: src/lib/export-helper.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
+function hasExportPermission (userPermissions, targetPermission) {
+  return userPermissions.includes(targetPermission)
+}
+
+function tooManyItems (resultCount, maxItems) {
+  return resultCount >= maxItems
```
Comment (reupen) [inline src/lib/export-helper.js]: It was > rather than >= before, > looked right to me?

Comment (jim68000) [reply src/lib/export-helper.js]: Might need to change the wording then. As it says you need to have fewer than 5000 items, so 5000 is actually not acceptable. Also, I've just noticed it says 'less than 5000 items' which is grammatically incorrect as well. I will try and grab Anthony for clarification.

**Comment by jim68000** [comment]
I’m going to WIP this as I’ve realised I need to deal with multiple possible values for the message rather than two

File: src/lib/export-helper.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+function buildExportAction (queryString, userPermissions, exportOptions) {
+  if (!hasExportPermission(userPermissions, exportOptions.targetPermission)) {
+    return {
+      enabled: false,
+    }
+  }
+  return {
+    enabled: true,
+    buildMessage: (count) => buildExportMessage(count, exportOptions.maxItems),
+    url: exportOptions.urlFragment + '/export?' + queryString,
```
Comment (yeahfro) [inline src/lib/export-helper.js]: Here we could use string interpolation

File: src/lib/export-helper.js [hunk 4]
```diff
@@ -1,16 +1,25 @@
...
-function tooManyItems (resultCount, maxItems) {
-  return resultCount >= maxItems
+function invalidNumberOfItems (resultCount, maxItems) {
+  return resultCount === 0 || resultCount >= maxItems
 }
 
 function buildExportMessage (resultCount, exportOptions) {
-  if (tooManyItems(resultCount, exportOptions.maxItems)) {
-    return `Filter to less than ${exportOptions.maxItems} ${exportOptions.entityPlural} to download`
+  const entityPlural = get(exportOptions, 'entityPlural', exportOptions.entityName + 's')
```
Comment (yeahfro) [inline src/lib/export-helper.js]: Looks like this `entityPlural` will never be set and the `+ 's'` will not work for `company`. We use `pluralise`.

Comment (jim68000) [reply src/lib/export-helper.js]: Actually, yes, I remembered pluralise on the way home last night.... the unset var here is me thinking ahead to the company/companies issue, but I will refactor to use pluralise instead

File: src/lib/export-helper.js (hunk 3 above)
Comment (jim68000) [reply src/lib/export-helper.js]: WIll change

**Pull request merged**